engine/execution/state/storehouse/queue/queue.go

```diff
@@ -0,0 +1,93 @@
+package queue
+
+import (
+	"sync"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type headerInContext struct {
+	Header *flow.Header
+	ID     flow.Identifier
+}
+
+// FinalizedBlockQueue is a concurrency-safe queue of finalized block headers
+// Every time a new header is added to the queue it verifies the complience of the header to the
+// most recenlty added header.
+type FinalizedBlockQueue struct {
+	headers          []headerInContext
+	lock             sync.RWMutex
+	expectedHeight   uint64
+	expectedParentID flow.Identifier
+}
+
+// NewFinalizedBlockQueue constructs a new FinalizedBlockQueue
+// "genesis“ is not inserted in the queue and is only used to to validate the very first incoming header
+func NewFinalizedBlockQueue(
+	genesis *flow.Header,
+) *FinalizedBlockQueue {
+	return &FinalizedBlockQueue{
+		headers:          make([]headerInContext, 0),
+		expectedHeight:   genesis.Height + 1,
+		expectedParentID: genesis.ID(),
+	}
+}
+
+// Enqueue checks the header compatibility and append a header to the queue
+// A header is compatible if its parentID matches the ID of the last inserted header (or genesis)
+// and its height is set right (last block's height plus one).
+// an error is returned if header not compatible
+func (ft *FinalizedBlockQueue) Enqueue(header *flow.Header) error {
+	ft.lock.Lock()
+	defer ft.lock.Unlock()
+
+	if ft.expectedHeight != header.Height || ft.expectedParentID != header.ParentID {
+		return &NonCompliantHeaderError{
+			ft.expectedHeight,
+			header.Height,
+			ft.expectedParentID,
+			header.ParentID,
+		}
+	}
+
+	blockID := header.ID()
+	ft.headers = append(ft.headers, headerInContext{header, blockID})
+	ft.expectedParentID = blockID
+	ft.expectedHeight = header.Height + 1
+
+	return nil
+}
+
+// Peek returns the oldest header in the queue without removing it
+// If the queue is empty it returns zero ID and nil header
+func (ft *FinalizedBlockQueue) Peek() (flow.Identifier, *flow.Header) {
+	ft.lock.RLock()
+	defer ft.lock.RUnlock()
+
+	if len(ft.headers) == 0 {
+		return flow.ZeroID, nil
+	}
+
+	header := ft.headers[0]
+	return header.ID, header.Header
+}
+
+// HasHeaders returns true if the queue is not empty and
+// has more headers to be consumed
+func (ft *FinalizedBlockQueue) HasHeaders() bool {
+	ft.lock.RLock()
+	defer ft.lock.RUnlock()
+
+	return len(ft.headers) > 0
+}
+
+// Dequeue removes the oldest header from the queue (without returning it)
+// Dequeue from empty queue is a no-op
+func (ft *FinalizedBlockQueue) Dequeue() {
+	ft.lock.Lock()
+	defer ft.lock.Unlock()
+
+	if len(ft.headers) > 0 {
+		ft.headers = ft.headers[1:]
```

Comment:
Probably want to overwrite first element and resize `headers` to:
- prevent underlying array of `headers` from growing continuously 
- garbage collect dequeued header 

Maybe something like:
```suggestion
		copy(ft.headers, ft.headers[1:])
		ft.headers[len(ft.headers)-1] = headerInContext{}
		ft.headers = ft.headers[:len(ft.headers)-1]
```

Reply:
you don't need to explicitly copy.  the headers slice will at most use 2x the array capacity.  once the currently allocated slice array rans out of capacity, the live data is copy over into a new array, thus freeing old/dequeued headers.

Reply:
> you don't need to explicitly copy. the headers slice will at most use 2x the array capacity. once the currently allocated slice array rans out of capacity, the live data is copy over into a new array, thus freeing old/dequeued headers.

You are right that it won't be growing continuously without explicit copy,  but `headers` will require new allocations continuously as we append to its end and reslice away its first element.  

Basically, I'm suggesting we can reuse instead of continuously allocating new underlying array.

Maybe for long-lived queue, it is better to reuse as much as possible (instead of allocating), and release dequeued header as early as possible (instead of waiting for next allocation).

Here's a program and its output (using simple byte slice) to to show initial capacity stays at 3 when using copy.

<details>
<summary> :computer:  Example program (click to expand)</summary>

```Go
package main

import "fmt"

func without() {
	b := []byte{1, 2, 3}

	b = b[1:]

	b = append(b, 4)
	fmt.Printf("capacity=%d, b=%v\n", cap(b), b)
}

func with() {
	b := []byte{1, 2, 3}

	copy(b, b[1:])
	b[len(b)-1] = 0
	b = b[:len(b)-1]

	b = append(b, 4)
	fmt.Printf("capacity=%d, b=%v\n", cap(b), b)
}

func main() {
	fmt.Printf("Without copy...\n")
	without()

	fmt.Printf("\nWith copy...\n")
	with()
}

```

</details>

Output:
```
Without copy...
capacity=8, b=[2 3 4]

With copy...
capacity=3, b=[2 3 4]
```

[Example program at Go playground](https://go.dev/play/p/YBU8cwmlg4l).